account_move_total_by_account_internal_group/hooks.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2022 ForgeFlow S.L.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+
+
+def pre_init_hook(cr):
+    """Precreate account_internal_group and fill with appropriate values to prevent
+    a MemoryError when the ORM attempts to call its compute method on a large
+    amount of preexisting moves."""
+    logger = logging.getLogger(__name__)
+    logger.info(
+        "Add account_move_line.account_internal_group column if it does not yet exist"
+    )
+    cr.execute(
+        "ALTER TABLE account_move_line ADD COLUMN IF NOT EXISTS account_internal_group VARCHAR"
+    )
+    cr.execute(
+        """ UPDATE account_move_line aml0 SET account_internal_group = aa.internal_group
+        FROM account_move_line aml
+        INNER JOIN account_account aa ON aa.id = aml.account_id
+        WHERE aml.id = aml0.id
+        AND aml.account_internal_group IS NULL
```

Comment:
I believe that you cannot use an `INNER JOIN` on an `UPDATE` clause. You need to use a `WITH` statement to create a temporary table to then use it on the `FROM`.

Reply:
This works, and it worked in a production database, because the FROM account_move_line acts like the WITH subquery in this case.

Can you show an example with the with query? I'd like to check what of the 2 queries is faster :)

Reply:
Oh, I thought it was not possible to use an INNER JOIN clause on the UPDATE. :face_with_spiral_eyes: I thought you always needed to use the WITH. Thank you! :smile: 

Yeah sure, here's an example:

```
WITH query as (
    SELECT id AS carrier_id, tariff_carrier_company_id AS tariff_carrier_id
    FROM delivery_carrier
    WHERE delivery_carrier_type = 'carrier'
    AND tariff_carrier_company_id IS NOT NULL
    GROUP BY id
)
UPDATE delivery_carrier dc
SET carrier_company_id = query.tariff_carrier_id
FROM query
WHERE dc.id = query.carrier_id;
```

Reply:
Here are the results, for a database of 791795 account move lines. The results of both queries are identical:
```
count	account_internal_group
473665	asset
17	equity
270358	expense
16517	income
31238	liability
```

Query 1: the same as it is in the init hook now:
EXPLAIN ANALYZE Query 1
```
Update on account_move_line aml0  (cost=0.57..81074.23 rows=3987 width=431) (actual time=62708.169..62708.170 rows=0 loops=1)
Planning Time: 36.767 ms
Execution Time: 62710.797 ms

```

Query 2:

```

WITH query as (
    SELECT aml.id as aml_id, aa.internal_group
    FROM account_move_line aml
    INNER JOIN account_account aa ON aa.id = aml.account_id
)
UPDATE account_move_line aml
SET account_internal_group = query.internal_group
FROM query
WHERE aml.id = query.aml_id;

```
EXPLAIN ANALYZE query 2:
```
Update on account_move_line aml  (cost=68170.26..130625.53 rows=792960 width=436) (actual time=65858.657..65858.659 rows=0 loops=1)
Functions: 18
Timing: Generation 22.297 ms, Inlining 0.000 ms, Optimization 9.456 ms, Emission 46.986 ms, Total 78.739 ms
Execution Time: 66050.844 ms
```

There is not much difference, but it seems Query 1 wins this time :)